src/t_hash.c

```diff
@@ -461,7 +471,7 @@ int hashTypeSet(robj *o, sds field, sds value, int flags) {
         hashtable *ht = o->ptr;
 
         sds v;
-        if (flags & HASH_SET_TAKE_VALUE) {
+        if (flags & HASH_SET_TAKE_VALUE || sdsType(value) == SDS_TYPE_32_SHARED) {
```

Comment:
Let's keep the semantics of TAKE_VALUE even for the shared sds:

* If HASH_SET_TAKE_VALUE is set, take over the ownership without incrementing the reference counter.
* Otherwise, increment the reference counter. This corresponds to a copy.